character-controller.js

```diff
@@ -456,130 +550,88 @@ class PlayerBase extends THREE.Object3D {
         }
       };
       _deinitPhysics();
-      
+
       const _removeApp = () => {
         this.removeActionIndex(wearActionIndex);
-        
         if (this.appManager.hasTrackedApp(app.instanceId)) {
           if (destroy) {
             this.appManager.removeApp(app);
-            app.destroy();
           } else {
             this.appManager.transplantApp(app, world.appManager);
           }
         } else {
-          // console.warn('need to transplant unowned app', app, this.appManager, world.appManager);
-          // debugger;
+          console.warn("need to transplant unowned app", app, this.appManager, world.appManager);
         }
       };
       _removeApp();
-      
-      const _emitEvents = () => {
-        app.dispatchEvent({
-          type: 'wearupdate',
-          player: this,
-          wear: false,
-          loadoutIndex,
-        });
-        this.dispatchEvent({
-          type: 'wearupdate',
-          app,
-          wear: false,
-          loadoutIndex,
-        });
-      };
-      _emitEvents();
+      this.handleWearUpdate(app, false, loadoutIndex, true, true)
+    } else {
+      this.handleWearUpdate(app, false, -1, true, true)
     }
   }
-  destroy() {
-    // nothing
-  }
-}
-const controlActionTypes = [
-  'jump',
-  'crouch',
-  'fly',
-  'sit',
-];
-class StatePlayer extends PlayerBase {
-  constructor({
-    playerId = makeId(5),
-    playersArray = new Z.Doc().getArray(playersMapName),
-  } = {}) {
-    super();
-
-    this.playerId = playerId;
-    this.playersArray = null;
-    this.playerMap = null;
-    this.microphoneMediaStream = null;
-    
-    this.avatarEpoch = 0;
-    this.syncAvatarCancelFn = null;
-    this.unbindFns = [];
-    
-    this.bindState(playersArray);
-  }
   isBound() {
     return !!this.playersArray;
   }
   unbindState() {
-    if (this.isBound()) {
-      this.playersArray = null;
-      this.playerMap = null;
+    this.playersArray = null;
+    this.playerMap = null;
+    if (!this.unbindFns) return;
+    for (const unbindFn of this.unbindFns) {
+      unbindFn();
     }
+    this.unbindFns.length = 0;
   }
-  detachState() {
-    throw new Error('called abstract method');
-  }
-  attachState(oldState) {
-    throw new Error('called abstract method');
-  }
-  bindCommonObservers() {
+  bindState(nextPlayersArray) {
+    // latch old state
+    const oldState = this.detachState();
+
+    // unbind
+    this.unbindState();
+    this.appManager.unbindState();
+
+    this.playersArray = nextPlayersArray;
+    if (!this.playersArray) return console.warn("this.playersArray is null")

```

Comment:
Probably a case that needs to be handled, depending on how the network state API is abstracted.